pubsub.go

```diff
@@ -942,10 +964,23 @@ func (p *PubSub) pushMsg(msg *Message) {
 	}
 
 	// reject unsigned messages when strict before we even process the id
-	if p.signStrict && msg.Signature == nil {
-		log.Debugf("dropping unsigned message from %s", src)
-		p.tracer.RejectMessage(msg, rejectMissingSignature)
-		return
+	if p.signPolicy.mustVerify() {
+		if p.signPolicy.mustSign() {
+			if msg.Signature == nil {
+				log.Debugf("dropping unsigned message from %s", src)
+				p.tracer.RejectMessage(msg, rejectMissingSignature)
+				return
+			}
+			// Actual signature verification happens in the validation pipeline,
+			// after checking if the message was already seen or not,
+			// to avoid unnecessary signature verification processing-cost.
+		} else {
+			if msg.Signature != nil {
+				log.Debugf("dropping message with unexpected signature from %s", src)
+				p.tracer.RejectMessage(msg, rejectUnexpectedSignature)
```

Comment:
We also need to handle this case in the score tracer, so that the peer is penalized but the message is not tracked -- see https://github.com/libp2p/go-libp2p-pubsub/blob/master/score.go#L574